xarray/tests/test_coding.py

```diff
@@ -50,3 +50,42 @@ def test_scaling_converts_to_float32(dtype):
     roundtripped = coder.decode(encoded)
     assert_identical(original, roundtripped)
     assert roundtripped.dtype == np.float32
+
+
+@pytest.mark.parametrize('dtype', 'u1 u2 i1 i2 f2 f4'.split())
+@pytest.mark.parametrize('scale_factor', [10, 0.01,
+                                          np.float16(10),
+                                          np.float32(10),
+                                          np.float64(10),
+                                          np.int8(10),
+                                          np.int16(10), np.int32(10),
+                                          np.int64(10), np.uint8(10),
+                                          np.uint16(10), np.uint32(10),
+                                          np.uint64(10), np.uint64(10)])
+@pytest.mark.parametrize('add_offset', [10, 0.01,
+                                        np.float16(10),
+                                        np.float32(10),
+                                        np.float64(10),
+                                        np.int8(10),
+                                        np.int16(10), np.int32(10),
+                                        np.int64(10), np.uint8(10),
+                                        np.uint16(10), np.uint32(10),
+                                        np.uint64(10), np.uint64(10)])
+def test_scaling_according_to_cf_convention(dtype, scale_factor, add_offset):
```

Comment:
This is a great set of invariant checks! But it would still be nice to see a more specific list of unit tests verifying that the decoded dtype is exactly as expected for a set of known inputs, e.g.,
```python
@pytest.mark.parametrize(
    ('input_dtype', 'scale_factor', 'add_offset', 'expected_dtype'),
    [
        (np.float32, 1, 0, np.float32),
        (np.float32, 1, 10, np.float64),
        ...
    ]
)
def test_cfscaleoffset_decoding_dtype(input_dtype, scale_factor, add_offset, expected_dtype):
    original = xr.Variable(...)
    decoded = variables.CFScaleOffsetCoder().decode(original)
    assert decoded.dtype == expected_dtype
```
This would us confidence that it's handling all the edge cases properly, e.g., only giving a bigger dtype when truly warranted.